text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,659 @@
+  - [System-specific extensions](#system-specific-extensions)
+  - [Examples](#examples)
+* [Future possibilities](#future-possibilities)
+  - [Transport context propagation](#transport-context-propagation)
+  - [Standards for context propagation](#standards-for-context-propagation)
+
+## Motivation
+
+This OTEP serves as a foundation for a first stable version of [messaging semantic conventions for tracing](https://github.com/open-telemetry/opentelemetry-specification/blob/a1a8676a43dce6a4e447f65518aef8e98784306c/specification/trace/semantic_conventions/messaging.md).
+It aims to define clear, consistent, and extensible conventions, it also
+describes reasons and motivations that led to the formulation of those
+conventions. The conventions comprise areas such as context propagation, trace
+structure, names, and attributes.
+
+After this OTEP is merged, the changes it proposes will be merged into the
+messaging semantic conventions, which will subsequently be declared
+[stable](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#stable).
+
+## Proposed stable messaging semantic conventions for tracing
+
+### Terminology
+
+The terminology used in this document is based on the [CloudEvents specification](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md).
+CloudEvents is hosted by the CNCF and provides a specification for describing
+event data in common formats to provide interoperability across services,
+platforms and systems.
+
+#### Message
+
+A "message" is a transport envelope for the transfer of information. The
+information is a combination of a payload and metadata. Metadata can be
+directed at consumers or intermediaries on the message path. Messages are
+transferred via one or more intermediaries.  Messages are uniquely
+identifiable.
+
+In the strict sense, a _message_ is a payload that is sent to a specific
+destination, whereas an _event_ is a signal emitted by a component upon
+reaching a given state. This document is agnostic of those differences and uses
+the term "message" in a wider sense to cover both concepts.
+
+#### Producer
+
+The "producer" is a specific instance, process or device that creates and
+publishes a message. "Publishing" is the process of sending a message or batch
+to the intermediary.
+
+#### Consumer
+
+A "consumer" receives the message and acts upon it. It uses the context and
+data to execute some logic, which might lead to the occurrence of new messages.
+
+The consumer receives, processes, and settles a message.
+
+* "Receiving" is the process of obtaining a message from the intermediary.
+* "Processing" is the process of acting on the information a message contains.
+* "Settling" is the process where intermediary and consumer agree on the state
+  of the transfer.
+
+#### Intermediary
+
+An "intermediary" receives a message for the purpose of forwarding it to the
+next receiver, which might be another intermediary or a consumer.
```

Comment:
I wonder if a peer-to-peer messaging system falls outside the scope of these conventions. Existence of an intermediary seems to be a base assumption here.

Reply:
Yeah I wonder the same. I remember a similar discussion when the concept of [channels](https://en.wikipedia.org/wiki/Channel_(programming)#Go) was brought up and if these conventions should also model that. Could be a v2 topic maybe?

Reply:
Actually, AFAIU from this "proposal", the intermediary doesn't really play an active role from a tracing point of view, right?
I mean, there are no spans that the intermediary is going to add on the trace even because it's mentioned here that it won't touch anything about tracing coming through producer and consumer but just forwarding messages. At same time, you cannot have control on the intermediary in the sense that not all messaging systems would implement tracing by adding spans and so on (i.e. Kafka itself doesn't have such a support). The intermediary mentioned in this proposal seems to have a role "just" from a messaging point of view not interfering with tracing at all. Why then it cannot be considered optional? Where producer sends directly to a consumer, the consumer settle to the producer, ...
Maybe I am missing some obvious tracing aspect and I don't see?

Reply:
This was discussed some time ago: https://github.com/open-telemetry/oteps/pull/173#discussion_r729120960

> I wonder if a peer-to-peer messaging system falls outside the scope of these conventions. Existence of an intermediary seems to be a base assumption here.

Yes. If the requirement for an intermediary is removed, there'd be a _wide_ range of use cases that might qualify as messaging. Those use cases should be supported, but I think they different enough to merit a different set of semantic conventions. 

> I mean ... with AMQP 1.0 protocol, the communication is really peer to peer so a producer could be connected directly to a consumer.

This is a case similar to [CloudEvents](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/cloudevents.md#overview). Both can be used for messaging use cases (with an intermediary) and for peer-to-peer communication (without intermediary).

I think distinguishing between messaging systems (requiring an intermediary) and peer-to-peer communication is crucial, as the presence of an intermediary and the asynchronous nature of communication between producer and consumer accounts for the special nature of messaging scenarios.

If the requirement for an intermediary is removed, it would be hard to argue why HTTP client/server communication wouldn't qualify as peer-to-peer messaging system.

Reply:
From a tracing stand point, the intermediary is not playing any role in a trace, you are right. This was a decision we took for a V1 where we would only focus on the "application layer" of the trace and not transport level. We have a project board, and you can see "intermediary instrumentation" is under "v2". https://github.com/orgs/open-telemetry/projects/20/views/1?filterQuery=v2

Reply:
> My point was raised just because I don't see the intermediary being in the "tracing" picture but we are assuming it's needed but in the end not that much for tracing purposes but for reducing the scope of the specification, which could make sense.

It was decided to define intermediary instrumentation as [out-of-scope for v1.0](https://github.com/open-telemetry/oteps/blob/main/text/trace/0173-messaging-semantic-conventions.md#instrumenting-intermediaries) of the semantic conventions. The first stable version intends to only cover producer and consumer instrumentation.

However, it is intended to cover intermediary instrumentation with future iterations of the semantic conventions. We definitely have the goal of keeping intermediary instrumentation in mind and to not come up with any conventions that block possibilities for intermediary instrumentation.

Also see [here](https://github.com/open-telemetry/oteps/blob/main/text/trace/0205-messaging-semantic-conventions-context-propagation.md#transport-context-propagation).

Reply:
Thanks guys!